RATEST-77: Attachments automated tests

kdaud opened the pull request:
Automated attachment feature in Reference Application
Ticket ID : https://issues.openmrs.org/browse/RATEST-77

HerbertYiga commented on src/test/java/org/openmrs/contrib/qaframework/RunTest.java:
```diff
@@ -1,8 +1,9 @@
 package org.openmrs.contrib.qaframework;
 
+import org.junit.runner.RunWith;
```
is this needed?

HerbertYiga commented on src/test/java/org/openmrs/contrib/qaframework/RunTest.java:
```diff
@@ -13,5 +14,6 @@
 		"html:target/index.html", "message:target/cucumber.ndjson"}, monochrome = true,
 // only scenarios tagged here are executed
 tags = "@selenium")
+
```
it looks like we don't need any changes from this file

kdaud replied:
Some things got messed up during conflict resolution. Am working out the best way and soon to address this

kaweesi replied:
you can clear the mess

kdaud replied:
Have addressed them and committed the changes

kaweesi replied:
i can still see some formatting changes coming in

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/automation/AttachmentSteps.java:
```diff
@@ -0,0 +1,117 @@
+	
+	@Given("a user logs in, searches a patient and visits Attachment page")
+	public void loadPatientDashboard() {
+		goToLoginPage().login(testProperties.getUsername(), testProperties.getPassword(), testProperties.getLocation());
```
you could make use of configuredUserLogin() here

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/automation/AttachmentSteps.java:
```diff
@@ -0,0 +1,117 @@
+		quit();
+	}
+	
+	@Given("a user logs in, searches a patient and visits Attachment page")
```
i don't see your 'is your and visits Attachment page' catered for in the block. and i can see the next step doing that

kdaud replied:
Have written the lines that cater for "visit Attachment Page" in the block

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/automation/AttachmentSteps.java:
```diff
@@ -0,0 +1,117 @@
+	@And("a user clicks on Attachment link")
+	public void userClicksOnAttachmentLink() {
+		attachmentsPage = new AttachmentsPage(dashboardPage);
+		attachmentsPage.clickOnAttachmentLink();
```
is the attachtmentLink clickable from the attachment page if its from the dashboard, we can move it to the right place

kdaud replied:
Moved the action to the right block

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/automation/AttachmentSteps.java:
```diff
@@ -0,0 +1,117 @@
+	
+	@And("a user attaches a file")
+	public void userAttachesFile() {
+		attachmentsPage.clickOrDropFile();
```
clickOrDropFile may be containing some slight typo!

kdaud replied:
Oops! It was an oversight, however,  I caught the typo after taking a look at it carefully

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/automation/AttachmentSteps.java:
```diff
@@ -0,0 +1,117 @@
+	@Then("the system enables capturing an image")
+	public void systemEnablesCapturingImage() {
+		attachmentsPage.clickOnCameraIcon();
+		assertNotNull(getElement(AttachmentsPage.ICON_CAMERA));
```
this does better coming first

kdaud replied:
Moved the block to run before the action

kaweesi replied:
i mean't the opposite, but if the clickon is happening for the same icon_camera element, you don't need to check it's existence since the clickon will fail if it doesn't exisit

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/page/AttachmentsPage.java:
```diff
@@ -0,0 +1,59 @@
+	}
+
+	public String addCaption() {
+		String captionText=userInput.next();
```
is there no better way to read the caption from the dom using selenium?

kdaud replied:
Applied sendKeys() to read the user input

kaweesi reviewed (commented):
Thanks @kdaud this is very promising, please find some comments from my review attached

kdaud commented:
Thanks @kaweesi  for reviewing my work and I feel motivated. I am working on the comments and ASAP making a commit with updated branch.

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/page/AttachmentsPage.java:
```diff
@@ -0,0 +1,55 @@
+
+public class AttachmentsPage extends Page{
+	
+	private static final By ATTACHMENTS_LINK = By.className("icon-share-alt edit-action right");
```
move this into the dashboard page within distro

kdaud replied:
Have refactored the code base

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/page/AttachmentsPage.java:
```diff
@@ -0,0 +1,55 @@
+	}
+
+	public void clickOnAttachmentAppLink() {
+		driver.findElement(ATTACHMENTS_LINK).click();
```
move this into the dashboard page within distro

kdaud replied:
> move this into the dashboard page within distro

 I understand referencing this element in my definition steps  is accessed through a library after moving it to the distro. Do I need to raise a PR to the `distro repo` since this feature does not exist in the upstream master. 
cc: @kaweesi

kaweesi commented:
Yeah, for the qa branch and you build the same branch locally

On Fri, Feb 26, 2021 at 1:54 PM Kakumirizi Daud <notifications@github.com>
wrote:

> *@kdaud* commented on this pull request.
> ------------------------------
>
> In src/test/java/org/openmrs/contrib/qaframework/page/AttachmentsPage.java
> <https://github.com/openmrs/openmrs-contrib-qaframework/pull/16#discussion_r583554076>
> :
>
> > +	private static final By UPLOAD_FILE = By.className("confirm ng-binding");
> +	private static final By CLEAR_FORMS = By.className("ng-binding");
> +	public static final By ICON_CAMERA = By.className("icon-camera");
> +	public static final By SAVE_BUTTON = By.className("icon-ok");
> +	
> +	public AttachmentsPage(ClinicianFacingPatientDashboardPage clinicianFacingPatientDashboardPage) {
> +		super(clinicianFacingPatientDashboardPage);
> +	}
> +
> +	@Override
> +	public String getPageUrl() {
> +		return "/attachments/attachments.page";
> +	}
> +
> +	public void clickOnAttachmentAppLink() {
> +		driver.findElement(ATTACHMENTS_LINK).click();
>
> move this into the dashboard page within distro
>
> I understand referencing this element in my definition steps is accessed
> through a library after moving it to the distro. Do I need to raise a PR to
> the distro repo since this feature does not exist in the upstream master.
> cc: @kaweesi <https://github.com/kaweesi>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-contrib-qaframework/pull/16#discussion_r583554076>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADZGJ4TUJR4EL53SHTGUKGLTA54XRANCNFSM4YD3RIPQ>
> .
>


-- 
*Joseph Kaweesi*
*Senior Full Stack Engineer, QA Engineer, Fellowship Mentor*
*Tel*: +256791763663, S*kype*: kaweesi.joseph, *IRC*: k-joseph, *Github*:
k-joseph/kaweesi, website: kaweesi.github.io

kdaud commented:
> Yeah, for the qa branch and you build the same branch locally

Hey @kaweesi  am resolving out this. Its just my network is a bit slow in building but soon will rise a Pr

kdaud commented:
> Yeah, for the qa branch and you build the same branch locally
> […](#)
> On Fri, Feb 26, 2021 at 1:54 PM Kakumirizi Daud ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In src/test/java/org/openmrs/contrib/qaframework/page/AttachmentsPage.java <[#16 (comment)](https://github.com/openmrs/openmrs-contrib-qaframework/pull/16#discussion_r583554076)> : > + private static final By UPLOAD_FILE = By.className("confirm ng-binding"); + private static final By CLEAR_FORMS = By.className("ng-binding"); + public static final By ICON_CAMERA = By.className("icon-camera"); + public static final By SAVE_BUTTON = By.className("icon-ok"); + + public AttachmentsPage(ClinicianFacingPatientDashboardPage clinicianFacingPatientDashboardPage) { + super(clinicianFacingPatientDashboardPage); + } + + @override + public String getPageUrl() { + return "/attachments/attachments.page"; + } + + public void clickOnAttachmentAppLink() { + driver.findElement(ATTACHMENTS_LINK).click(); move this into the dashboard page within distro I understand referencing this element in my definition steps is accessed through a library after moving it to the distro. Do I need to raise a PR to the distro repo since this feature does not exist in the upstream master. cc: @kaweesi <https://github.com/kaweesi> — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#16 (comment)](https://github.com/openmrs/openmrs-contrib-qaframework/pull/16#discussion_r583554076)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADZGJ4TUJR4EL53SHTGUKGLTA54XRANCNFSM4YD3RIPQ> .
> -- *Joseph Kaweesi* *Senior Full Stack Engineer, QA Engineer, Fellowship Mentor* *Tel*: +256791763663, S*kype*: kaweesi.joseph, *IRC*: k-joseph, *Github*: k-joseph/kaweesi, website: kaweesi.github.io

I have

kdaud commented:
I have refactored the code and made a [PR](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/300) on the openmrs-distro

kdaud commented:
I think the better logic(implementation) of attachment page as a resource to the definition steps  would be  made in the distro module rather than in the qaframework module. The resource can then be accessed through a library. 
I have no objection to refactor the work in case it makes the better implementation. @kaweesi  Any  thought on this

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/automation/AttachmentSteps.java:
```diff
@@ -0,0 +1,115 @@
+	
+	@And("a user enters a caption")
+	public void userEntersCaption() {
+		attachmentsPage.addCaption("'");
```
is there any reason for using ' as your caption?

kaweesi commented on src/test/resources/features/15-attachment/attachments.feature:
```diff
@@ -0,0 +1,29 @@
+ Feature: Attachments Management
```
Please test this locally, i have tried doing this and all the 3 scenarios were failing

```
Tests in error: 
  Uploading Attachment(Attachments Management)
  Removing un intended Attachment(Attachments Management)
  Capturing an image(Attachments Management)

```

kdaud replied:
Hey @kaweesi is there a way @CucumberOptions can allow running selenium test against a single file and ignore other files present in the same folder. After making a research on this, glue option can be used to define the definition step package.  
Also i would like to test the feature under construction on my local instance of standalone. Is this possible i can see in the test.properties file the web-url redirects at the remote server. If so is it ok to do some changes in this file to meet my requirement. Thanks

kaweesi commented:
Please tag your special test you want to run exclusively without running
others and update
https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/src/test/java/org/openmrs/contrib/qaframework/RunTest.java#L15
to use your new tag.

You can change settings in
https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/src/test/resources/org/openmrs/uitestframework/test.properties
to your local instance etc.

But remember none of those 2 above local changes need to get pushed since
that would interfere with our ci builds.

Always use git
checkout src/test/resources/org/openmrs/uitestframework/test.properties and git
checkout src/test/java/org/openmrs/contrib/qaframework/RunTest.java to
reset the 2 files once you are done with your tests and wants to commit
before pushing

On Mon, Mar 1, 2021 at 6:12 PM Kakumirizi Daud <notifications@github.com>
wrote:

> *@kdaud* commented on this pull request.
> ------------------------------
>
> In src/test/resources/features/15-attachment/attachments.feature
> <https://github.com/openmrs/openmrs-contrib-qaframework/pull/16#discussion_r584792434>
> :
>
> > @@ -0,0 +1,29 @@
> + Feature: Attachments Management
>
> Hey @kaweesi <https://github.com/kaweesi> is there a way @CucumberOptions
> can allow running selenium test against a single file and ignore other
> files present in the same folder. After making a research on this, glue
> option can be used to define the definition step package.
> Also i would like to test the feature under construction on my local
> instance of standalone. Is this possible i can see in the test.properties
> file the web-url redirects at the remote server. If so is it ok to do some
> changes in this file to meet my requirement. Thanks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-contrib-qaframework/pull/16#discussion_r584792434>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADZGJ4VLOKCP7ULN2WCTKW3TBOVHNANCNFSM4YD3RIPQ>
> .
>


-- 
*Joseph Kaweesi*
*Senior Full Stack Engineer, QA Engineer, Fellowship Mentor*
*Tel*: +256791763663, S*kype*: kaweesi.joseph, *IRC*: k-joseph, *Github*:
k-joseph/kaweesi, website: kaweesi.github.io

kdaud commented:
Thanks for your time, i feel have bothered you a lot but please allow me do so, its worth it. You have made me now know what to do in regard to local testing.

kaweesi commented:
no worries at all @kdaud , i am here to support you as much as there's need in your career development. So you are invited to bother me as much as you can to grow your qa and dev expertise

kdaud commented:
> Also remember to change back your tag to @Selenium in your feature file in addition to resetting the above 2 files

Ideally does it mean when doing my stuff locally, have to tag the scenarios under the test with @run and after meeting the requirements then refactor it to @Selenium for commit

kaweesi commented:
that can be ok where you want to run just your test, once you reset your test change, remember to test again in-case your change affected any other existing tests

kdaud commented:
> that can be ok when you want to run just your test, once you reset your test change, remember to test again in-case your change affected any other existing tests

Cool!

kdaud commented:
Hey @kaweesi  have been able to reproduce the bug after overcoming the blocker. Am fixing some failing test and will be able to update the pr soon

kdaud commented:
Have spent some good hours while debugging failing tests. However, have realized that some elements could not be retrieved by selenium driver from the page. Most likely it a result of some page elements sharing identical attributes. So am now testing different element attributes(id, className, cssSelector, etc) that can uniquely identify these elements. @k-joseph which attribute could be suitable to uniquely identify an element on a page otherwise it has taken me almost 4 hours to identify this!

kaweesi commented:
use chrome tools to inspect the element and then consider using the most unique way you can think about considering the element attributes and use that instead.

sherrif10 commented:
could  you please resolve merge conflicts

kdaud commented:
CI is unable locate the page resource refactored to distro. However, have made a [commit|https://github.com/openmrs/openmrs-distro-referenceapplication/pull/328] to the distro

sherrif10 commented:
@kaweesi  recently commited a fix for travis, could you try to update this pull request to the latest changes

kdaud commented:
> @kaweesi recently commited a fix for travis, could you try to update this pull request to the latest changes

Have you had

kdaud commented:
Am working on the recommendation of the reviews and soon to update this commit. However, from the technical perspective  travis will always fail or [is mostly likely fail] on tickets of this kind because of the tests resources are coded in the distro. On the trigger of looking for this resources in distro, travis will find null values present and by default will fail @kaweesi could this suggest merging the resources first before raising the tests pr which is not a good idea because the committer may require to make some changes during test reviews. Any thought on this @sherrif10 @HerbertYiga

kaweesi commented:
It's still ok working on them together, but we can only review and merge
the dependency first and then if a dependeded PR exists should trigger
travis once again. I hope we can have the dependency resolved first @kdaud

On Wed, Mar 17, 2021 at 5:20 PM Kakumirizi Daud ***@***.***>
wrote:

> Am working on the recommendation of the reviews and soon to update this
> commit. However, from the technical perspective travis will always fail or
> [is mostly likely fail] on tickets of this kind because of the tests
> resources are coded in the distro. On the trigger of looking for this
> resources in distro, travis will find null values present and by default
> will fail @kaweesi <https://github.com/kaweesi> could this suggest
> merging the resources first before raising the tests pr which is not a good
> idea because the committer may require to make some changes during test
> reviews. Any thought on this @sherrif10 <https://github.com/sherrif10>
> @HerbertYiga <https://github.com/HerbertYiga>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-contrib-qaframework/pull/16#issuecomment-801119674>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADZGJ4TTBUBTGWMW7WUD6KDTEC3B5ANCNFSM4YD3RIPQ>
> .
>


-- 
*Joseph Kaweesi*
*Senior Full Stack Engineer, QA Engineer, Fellowship Mentor*
*Tel*: +256791763663, S*kype*: kaweesi.joseph, *IRC*: k-joseph, *Github*:
k-joseph/kaweesi, website: kaweesi.github.io

HerbertYiga commented:
> It's still ok working on them together, but we can only review and merge the dependency first and then if a dependeded PR exists should trigger travis once again. I hope we can have the dependency resolved first

Was coming to say the same

kdaud commented:
Here is the resource [commit](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/328) to this test

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/automation/AttachmentSteps.java:
```diff
@@ -0,0 +1,120 @@
+	@Before(RunTest.HOOK.SELENIUM_DASHBOARD)
+	@Override
+	public void initiatePatientDashboard() {
+		initiateWithLogin();
```
you don't need this when you use the @Before(RunTest.HOOK.SELENIUM_DASHBOARD) hook

kdaud replied:
Alright !!

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/automation/AttachmentSteps.java:
```diff
@@ -0,0 +1,120 @@
+		initiateWithLogin();
+		findPatientPage = (FindPatientPage) homePage.goToFindPatientRecord().waitForPage();
+		findPatientPage.clickOnFirstPatient();
+		dashboardPage = (ClinicianFacingPatientDashboardPage) findPatientPage.clickOnFirstPatient().waitForPage();
```
you also don't need all these at all. just include one call; `initiatePatientDashboard();` consider forexample

https://github.com/openmrs/openmrs-contrib-qaframework/blob/master/src/test/java/org/openmrs/contrib/qaframework/automation/AllergiesSteps.java#L19

kdaud replied:
> just include one call; `initiatePatientDashboard();`

Ooh yes. Thanks

kaweesi commented on src/test/java/org/openmrs/contrib/qaframework/automation/AttachmentSteps.java:
```diff
@@ -0,0 +1,120 @@
+			dashboardPage.startVisit();
+			attachmentsPage = (AttachmentsPage) dashboardPage.clickOnAttachmentsAppLink().waitForPage();
+		} 
+		else if (dashboardPage.containsText("End Visit")) {
```
refactor this part to only inculde one if to start the visit and the clicking on attatmentsPage out

kdaud replied:
Thanks @kaweesi . This is addressed in the updated pr

kdaud commented:
> Please also consider my reviews on the sister PR to this at: [openmrs/openmrs-distro-referenceapplication#328](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/328)

Thanks for the reviewing my work. Am addressing all the reviews and soon to update both pr's
